sound/soc/sof/loader.c

```diff
@@ -854,6 +854,10 @@ EXPORT_SYMBOL(snd_sof_run_firmware);
 
 void snd_sof_fw_unload(struct snd_sof_dev *sdev)
 {
-	/* TODO: support module unloading at runtime */
+	/* Don't cache a firmware that failed to boot.
+	 * TODO: fully support module unloading at runtime too.
+	 */
+	release_firmware(sdev->pdata->fw);
+	sdev->pdata->fw = NULL;
```

Comment:
in terms of what it achieves today , this is fine. But when and if  we do start supporting module unload at runtime, wot's this need to be moved before snd_sof_remove()?

Reply:
I don't understand your comment sorry. Do you mean `release_firmware()` should be invoked before `snd_sof_remove()` in the future? I don't see why. As @ujfalusi  wrote, the only reason we need the `struct firmware` again is to resume after suspend (especially on platforms without IMR resume), even for that it shouldn't be needed eventually. So releasing `struct firmware` before or after some releasing other things does not matter, does it? Of course the sooner the better but I don't see any required order or dependency.


I asked the compiler to list all places where we use `struct firmware`, here is the complete list as of commit 132b2eebcfcd
```
snd_sof_load_firmware_raw()
snd_sof_load_firmware_memcpy()
snd_sof_fw_unload()
hda_dsp_cl_boot_firmware()
hda_dsp_cl_boot_firmware_iccmax()
```
It's not used anywhere else in my `.config`.

Reply:
I'm not sure why the order of release_firmware() would matter, in `snd_sof_device_remove()` the firmware is not needed, we could have drop it first thing or last thing. It does not really matter.

@ranj063, what does "module unload at runtime" means? Do you want to be able to remove the modules while audio is playing or something else?
Are we preparing for hot-plugable DSPs?

Reply:
from what I understood, I think runtime module unload means you load/unload FW modules based on topology. But I could be wrong. And in any cae, we're a bit far from supporting this at the moment. So this PR is good to go for now.